File: backend/app/schemas/building.py
import uuid
from datetime import datetime
from typing import Any, Literal

from pydantic import BaseModel, Field


# ---------- Zone sub-models ----------

class DaySchedule(BaseModel):
    name: str
    start_month: int = Field(..., ge=1, le=12)
    start_day: int = Field(..., ge=1, le=31)
    end_month: int = Field(..., ge=1, le=12)
    end_day: int = Field(..., ge=1, le=31)
    days: list[int]    # 1=Mon … 7=Sun
    hours: list[int]   # 0-23
    value: float


class ParamConfig(BaseModel):
    mode: Literal["fixed", "scheduled"]
    fixed_value: float
    schedules: list[DaySchedule] = []


class WallConfig(BaseModel):
    """Per-wall exterior/interior setting."""
    south_exterior: bool = True
    north_exterior: bool = True
    east_exterior: bool = True
    west_exterior: bool = True


class BuildingZone(BaseModel):
    name: str = Field(..., max_length=50)
    area: float = Field(..., ge=0.1, le=9999.9)
    floor_height: float = Field(3.5, ge=1.0, le=100.0)
    # Zone vertical position: determines floor/roof boundary conditions
    zone_position: Literal["top", "middle", "bottom", "single"] = "single"
    # Wall exterior/interior config
    wall_config: WallConfig = WallConfig()
    # Envelope
    wall_u_value: float = 0.6
    window_u_value: float = 2.2
    window_wall_ratio: float = 0.3
    roof_u_value: float = 0.4
    # Internal gains
    people_density: ParamConfig
    lighting_density: ParamConfig
    equipment_density: ParamConfig
    fresh_air_volume: ParamConfig
    # Setpoints (optional for backward compat)
    temperature: ParamConfig = ParamConfig(mode="fixed", fixed_value=26, schedules=[])
    relative_humidity: ParamConfig = ParamConfig(mode="fixed", fixed_value=50, schedules=[])


# ---------- Building CRUD schemas ----------

class BuildingCreate(BaseModel):
    name: str = Field(..., max_length=200)
    building_type: str = Field(..., max_length=100)
    total_area: float | None = None
    floor_count: int | None = None
    location: str | None = Field(None, max_length=200)
    climate_zone: str | None = Field(None, max_length=50)
    envelope_params: dict[str, Any] | None = None
    zones: list[BuildingZone] | None = None


class BuildingUpdate(BaseModel):
    name: str | None = Field(None, max_length=200)
    building_type: str | None = Field(None, max_length=100)
    total_area: float | None = None
    floor_count: int | None = None
    location: str | None = Field(None, max_length=200)
    climate_zone: str | None = Field(None, max_length=50)
    envelope_params: dict[str, Any] | None = None
    zones: list[BuildingZone] | None = None


class BuildingResponse(BaseModel):
    id: uuid.UUID
    project_id: uuid.UUID
    name: str
    building_type: str
    total_area: float | None
    floor_count: int | None
    location: str | None
    climate_zone: str | None
    envelope_params: dict[str, Any] | None
    zones: list[BuildingZone] | None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}

File: frontend/src/types/building.ts
export interface DaySchedule {
  name: string          // e.g., "工作日白天", "周末"
  start_month: number   // 1-12
  start_day: number     // 1-31
  end_month: number     // 1-12
  end_day: number       // 1-31
  days: number[]        // 1=Mon, 2=Tue, ..., 7=Sun
  hours: number[]       // 0-23, which hours have this value
  value: number         // single parameter value
}

export interface ParamConfig {
  mode: 'fixed' | 'scheduled'
  fixed_value: number
  schedules: DaySchedule[]
}

/** Zone vertical position in the building */
export type ZonePosition = 'top' | 'middle' | 'bottom' | 'single'

/** Per-wall exterior/interior setting */
export interface WallConfig {
  south_exterior: boolean
  north_exterior: boolean
  east_exterior: boolean
  west_exterior: boolean
}

export interface BuildingZone {
  name: string
  area: number
  floor_height: number
  // Zone position (determines floor/roof boundary conditions)
  zone_position: ZonePosition
  // Wall exterior/interior config
  wall_config: WallConfig
  // Envelope
  wall_u_value: number
  window_u_value: number
  window_wall_ratio: number
  roof_u_value: number
  // Internal gains with schedule support
  people_density: ParamConfig
  lighting_density: ParamConfig
  equipment_density: ParamConfig
  fresh_air_volume: ParamConfig
  // Setpoints with schedule support
  temperature: ParamConfig
  relative_humidity: ParamConfig
}

// Keep for backward compatibility
export interface EnvelopeParams {
  wall_u_value?: number
  window_u_value?: number
  window_wall_ratio?: number
  roof_u_value?: number
  people_density?: number
  lighting_density?: number
  equipment_density?: number
  fresh_air_volume?: number
}

export interface Building {
  id: string
  project_id: string
  name: string
  building_type: string
  total_area: number | null
  floor_count: number | null
  location: string | null
  climate_zone: string | null
  envelope_params: EnvelopeParams | null
  zones: BuildingZone[] | null
  created_at: string
  updated_at: string
}

export interface BuildingCreate {
  name: string
  building_type: string
  total_area?: number
  floor_count?: number
  location?: string
  climate_zone?: string
  envelope_params?: EnvelopeParams
  zones?: BuildingZone[]
}

export interface BuildingUpdate {
  name?: string
  building_type?: string
  total_area?: number
  floor_count?: number
  location?: string
  climate_zone?: string
  envelope_params?: EnvelopeParams
  zones?: BuildingZone[]
}
